### ISSUE TITLE
Feature Request:  Collapse filters to help people navigate them

### ISSUE THREAD

**Issue by BendixSpring** [TeamMentor/TM_4_0_Design#350]
I think a twisty on each of the main filter headings would help.  Too much info on the screen.

Have the current filter opened, and everyone else collapsed.  Unfortunately, it appears that you can only have one filter.  Google has better search capabilities.

Collapse these:
Category
Technology
Phase
Type

I think you should break out "Language" from technology and create another filter category called "language".

**Comment by arvinddoraiswamy** [comment]
Agreed. Too much scrolling needed to find out what I'd like to filter by. I'd like only the Top level filters to be displayed.
Category, Tech, Phase, Type

Then only when I click on Category should a new expansion happen. Preloading everything is not logical IMO. Why would a user want to know about every category right at the start?

**Comment by luchia** [comment]
Hey guys, I did some work on this yesterday on #348 that I think addresses this problem. Let me know what you think.

![filters](https://cloud.githubusercontent.com/assets/4547737/5951148/a5b1608e-a7ae-11e4-8fa9-ff4458ed1abc.gif)
![screen shot 2015-01-29 at 12 49 49 pm](https://cloud.githubusercontent.com/assets/4547737/5951613/58a2ff30-a7b5-11e4-89ef-9448b22f9682.png)
![screen shot 2015-01-29 at 12 49 51 pm](https://cloud.githubusercontent.com/assets/4547737/5951615/58a5da7a-a7b5-11e4-9fee-995a19318b12.png)
![screen shot 2015-01-29 at 12 49 53 pm](https://cloud.githubusercontent.com/assets/4547737/5951614/58a4adda-a7b5-11e4-8645-b50eaf413af1.png)

**Comment by BendixSpring** [comment]
From what I see, this looks good.  In the JavaScript version, would we see multiple selections?

 

For example:

Under technology, I might select Java and PHP – would only they be listed?

Or if I selected something from Technology, Phase, and Type – would only those particular filters be listed?

 

If so, that would be great!

 

Thanks,

Danny

**Comment by luchia** [comment]
Hi Danny, it's difficult to explain and the workflow isn't ready to show yet as @DinisCruz hasn't built the JS for that feature. I can tell you that it is easy to use and shows multiple filter selections which load the screen dynamically.

**Comment by BendixSpring** [comment]
Great!  I’m looking forward to giving it a try!

**Comment by DinisCruz** [comment]
@BendixSpring the plan is to support multiple selections in Flare (the JS version of TM 4.0)

**Comment by BendixSpring** [comment]
Am I correct in assuming that if the system detects Java Script enabled in the browser, it will use Flare, otherwise the non-JS version?

**Comment by DinisCruz** [comment]
@BendixSpring yes that is what WILL eventually happen. 

At the moment everybody sees Jade (non-JS-version) since that is the version currently in a usable state

**Comment by arvinddoraiswamy** [comment]
Much much better. This makes so much more sense :)

**Comment by roman87** [comment]
@BendixSpring  In the java script version we will have the ability to select multiple filters. See the high level requirement that we are going for here https://github.com/TeamMentor/Master/issues/899
The challenge at that time will  be is to figure out what filters should be ANDed and what ORed and how to intuitively show this to the user. Also the idea is to have a lot more filters/tags for each article.

**Comment by BendixSpring** [comment]
This will be great!  

 

One possible way of doing the AND/OR would be to use something like this:

Match Any term (OR)

Match All terms (AND)

**Comment by roman87** [comment]
Not to to get into a lengthy discussion ahead of time :), but it is what terms can be Any (OR)  and what terms can be All (AND). i.e. should we say that I should be able to pick multiple technologies, AND only one phase? (T1 U T2 U T3)   ∩ Ph1 or should we allow multiple phases to be picked? (T1 U T2 U T3) ∩ (Ph1 U Ph2) or if article matches any technology and any selected phase T1 U T2 U Ph1 U P2. But what if the user wants an article that is for .NET 3.5 and for .NET 4.0  and ether P1 or P2 (T1 ∩ T2) ∩ (P1 U P2).. you see how quickly this gets complicated. Technically it is easy to do any of these and we may even provide some advanced syntax at some point for the user to do any combination. But Visually we need to be able to limit our UI to just the "right" combination of filters that they do not overwhelm the first time unsuspecting user like they do in 3.5 and they are functional enough to be useful.

**Comment by BendixSpring** [comment]
I’m in favor of a lengthy discussion if it will help us build a better product :)

 

That being said, I proposed a very simple radio button design.

 

However, to be truly useful to the power user, I’d like to be able to use Google-style operators to allow me to build complex queries as you described below.  (That will be especially important in the non-JS version, I think.)

 

So using the Google model, I’d like to see the following: 

AND (implicit, default)

OR

(+) must include

(-) ignore term

“exact phrase”

 

--Danny

**Comment by roman87** [comment]
I agree. Still the question is what filter get radio buttons and what filters get checkboxes :)
Anyways, I think we should build a couple of prototypes and see which one works best. And yes 100% onboard with the google type advanced filtering.

**Comment by arvinddoraiswamy** [comment]
If we do the above (complex filter building) the user must be helped through the regex building, not thrown awful errors like - "There was an error in your regular expression".

You don't say? What IS a regular expression? What a terrible product!!

Slightly exaggerated but you get the point :)

I think the Amazon "Product filter" model is a good example though. Very intuitive. NO think. Click. Click. Click. Would you guys agree? Maybe we could look there before building this.

**Comment by roman87** [comment]
Well.. there are two kinds of users for the filters. Ones that would like to go for the advanced features, and the ones that are very happy with the simplified features. We can provide all the click-click options for the advanced users we want, but if we don't get the simplified features right, we have a problem. Simply because MOST users will use the simplified features.

**Comment by arvinddoraiswamy** [comment]
Totally.

In which case... we should just:

- Do simple (which we have now, I assume?)
- See if it works (Can users use it without crying?)

if yes:
   look at Amazon
   s/simple/advanced click click/
   see if it works

repeat till bored :)

The end :)

**Comment by roman87** [comment]
The simple one is the challenge. I don't think what we have now works. I could be wrong :)

**Comment by arvinddoraiswamy** [comment]
Right so then... THAT is what we should talk about. Maybe you're rt.. maybe
you're wrong. But we should talk about it. Using the 5 WHYs appraoch that
Tyler (my manager :)) introduced me to....

Here's number 1 - Why do you not feel that what we have now doesn't work?

**Comment by roman87** [comment]
Haha.. well it does not work because it allow you to select only a single filter. As a result it does not provide (IMO) robust enough filtering to isolate an article the user is looking for.

**Comment by arvinddoraiswamy** [comment]
Great. Thanks for trying this appraoch out :). So then question 2 is...

Why does the UI allow the user to select only 1 filter?
    The obvious answer is... because the devs said so :)

So I'll skip to the next pertinent question...

Why did the TM team design it so the user could select only 1 filter? What
was the thought process behind it?

**Comment by roman87** [comment]
Haha... well your answer to the 2nd question is also the answer to the 3d. :).. no really, because this is a first beta version and this was the easiest to implement. What we need to figure out is what is it that we want to implement when we tackle a harder problem :) Which is what this thread is all about :)

**Comment by BendixSpring** [comment]
Arvind:

 

I do agree that the Amazon product filter design is worth considering.  I think does “ANDing” which may not be a bad default.

 

But the full power of searching should still be at the search box with operators.

**Comment by BendixSpring** [comment]
Agreed.  Broken by default, unfortunately.

**Comment by arvinddoraiswamy** [comment]
Hehe cool. No more whys then :) .. we got an answer by the 3rd one ;)

IMO and this is purely IMO... I'd wait till you get some client feedback
before you think of solving this problem. Because right now, maybe we dont
*have* a problem at all? Maybe clients will love this?

Coz I mean... there's no right answer - maybe we need more data before we
can decide either way?

If you thought this was the easiest to implement - maybe it is also because
it is truly the simplest way.. from a user thought process perspective. I'd
back that thinking now and hibernate this thread.

Again all totally IMO :)

**Comment by roman87** [comment]
See, the problem?! We have two power users that have two opinions :) That is why we need to nail the simplified search, so that you will rarely have to go to the power search.

**Comment by roman87** [comment]
And yes Arvind, I definitely agree that we need field trials. But we already have one power user @BendixSpring asking if we will have multiple selections in further versions :)

**Comment by roman87** [comment]
btw @arvinddoraiswamy  I am anxious to hear what are all the 5 questions :)?

**Comment by arvinddoraiswamy** [comment]
Right but then... we're back where we started :D

What IS simple?
- A user wants a article quickly.
- The user gets the article quickly.

That is simple.

If you ask me if filtering is the best *way* to do it...I'd say no. I'd
just keep a search box. Meaning the flow is:

- Log in
- Search
- Get article
- Logout

That is it. But now that's not a popular approach :). :sneaks away silently

**Comment by arvinddoraiswamy** [comment]
Its addictive I tell you and a wonderful problem solving approach :)

The logic is each time you say ... something .... anything... you question
its true motive. Here's an example (highly contrived and may or may not be
real ;)):

Arvind: I don't think the filter approach works
Roman: Why?
Arvind: Its too complex.
Roman: Why is it too complex?
Arvind: I don't know which filter to choose?
Roman: Why do you not know which filter to choose?
Arvind: There are too many filters.

At this point though... asking... *Why* are there too many filters? is the
wrong question.... its obviously coz the devs put them there. So we have
reached a conclusion for this set. And we ask a new set:

Arvind: Why did the devs put them there?
Roman: Dunno. Coz they copied them from the previous release?
Arvind: Why?
Roman: Coz we didn't have time to rethink :)

We could go down...WHY you didnt have time? But that's a tangent ... so
we'll stop it there :)

But now after these questions... Roman could think... "hey maybe the
filters were always broken?" ... maybe its important to think about them if
they are central to this release.

And NOW you go one of three ways:

- Look at what filters users most often clicked on in past releases and use
JUST those here
- Throw all the Filters out and design a new set, which goes back to....
BUT which ones.... so you need step 1 anyway :)

Then you find out... you have no user data... in which case... you come
back to Step 2..

- Throw all the Filters out and design a new set, which goes back to....
BUT which ones....

If Step 2 is still failing... it means you don't know or/and.....(mostly
and) don't have enough data to decide.


So.... you could...
-- still go ahead with a new set and hope it works?

Or scrap the filter method and use a better search mechanism :)

No I'm not *pushing* search here... just trying to explain my thought
process.

I hope its partially helpful atleast... specially coz I'm writing it at 1
am on a Sunday night :) :)

**Comment by roman87** [comment]
It is the approach. However, searching is not enough. Once the user gets a hang of the taging/keyword/filtering system it is much easier to search and then narrow down the search through these keywords/filters.

**Comment by arvinddoraiswamy** [comment]
Hmm I dont know about that :)

Think of it... if you wanted to learn how to grow pumpkins in your
garden...what would you type into Google? I'd type:



*How to grow pumpkins beginner*... learn from there... and then learn the
keywords.

I wouldn't do it the other way around. For me ... that is broken. That way
works .. for experienced users... not beginners. And I feel we should
assume all users are beginners here.

1 more... Why does Gmail work? coz its simple... and all i do is search...
*i* assign labels... *i* create folders... *after* knowing what I have in
my mail. Not before.

But hey... maybe I'm wrong and you're right :). Its an approach... that's
all... only way to know for sure is to ask users.

Either way - this is a fun conversation :). I hope you feel that way too.

**Comment by roman87** [comment]
Definitely a fun conv :)
So I just went to google and typed "how to grow pumpkins beginner"  I got about 8 million hits. Not going past the first page there are mainly sites like "pumpkinmaster.com" informeddemocracy.com/pumpking/growing  10 steps to a giant pumpkin.  So now, since I really know nothing about punkins, I would have to go and read through at least a few of these sites, figuring out which ones I can more or less trust, which ones have actual meaningful info and finally come up with some idea on how to grow a pumpkin.  
Now imagine going into a library, finding the gardening section, under there finding vegetables, and under letter P finding a How to grow pumpkin for dummies.  
Of' course, I had to know that pumpkin is a vegetable, it starts with letter P and that there is a gardening section in the library. I actually had to have a library card too :)
So what if we combine the best of both worlds? In a world of faceted search.

**Comment by arvinddoraiswamy** [comment]
TL;DR - I agree. Having both is cooler. I'll explain. There might be some
redesign...but I think it makes sense to *atleast *think about it. I feel
its the best way.

Nice :) :) we're getting closer.

"Of' course, I had to know that pumpkin is a vegetable, it starts with
letter P" - Correct

but knowing about gardening... maybe..maybe not. I could go to
vegetables... search for plant, grow and garden... and then get to garden.
Or maybe I'd get to garden immediately. Depends on the user. Meaning... if
I know the tag.... this approach is great. If not, and I have a ton of
tags... maybe not that great.

Meaning... if we have only a few tags.... this approach could work...even
in the worst case scenario where the user brute forces all "garden" verbs.
Less is more.

Meaning.... how can we have tags that the user will definitely know. 100%.
Even the most basic user. For that we go back to "who" the most basic user
is. What will she know?... at an absolute bare minimum. I think its one of
2 things:

- Technology name (.NET, Java, C)
- Vuln name (SQLi, XSS, CSRF)

They'll definitely know one of these 2... else lets face it  they wouldn't
have a job :). Meaning they will search based on either of these. Either...
by language or by vulnerability. And if we go deeper... vulnerability is a
subset of a language. A language always exists... a vuln may or may not.
Java IS there... buffer overflows in Java...aren't.

Which means... if you want to have a filter based approach... the only tag
should be languages. That's it. And each vuln under that...tree wise or
however you want the UI to be. So a user could search for "SQL Injection"
and find all languages that have this vuln...something like....

SQLi:
<link> Path to .NET SQLi </link>
<link> Path to Java SQLi </link>

Click.

Tree expands and takes the user there... so next time... the user doesn't
have to search :). Inside this SQLi article - have everything you currently
have under Type. Bookmark links ON the page so users can jump up and down
on the page. I have everything for SQLi on .NET on 1 page. Everything.

Sure you can search in a million other ways....but do you ..need to? When
you can do it with just..1 tree?

If the answer to that is...yes... why not let the user customize his tree
over time..himself? Let them create whatever hierarchy they want? We give
them the content - in 1 simple form we feel is the best. Let them.. tweak
it.

Coz lets face it.... folder structure is up there as one of the most
disagreed upon issues in technology :)

Anyway this was interesting enough to make me write an Email at 3 am ..
lol. I hope it helps and you guys "fix" it... ;)

**Comment by DinisCruz** [comment]
@arvinddoraiswamy I'm completely with your that the workflow for TM should be 

> - Log in
> - Search
> - Get article
> - Logout

(and even the Log in and Logout should be made easier with things like SSO and OAuth)

My view is that TM is a 'Search Driven' solution, where most users will want to 'search their way into a solution' (or in our case, an article relevant to their question)

Not that we shouldn't have a way to navigate the data, but the key to get right in TM is the search

Btw, thx for the feedback, and please keep coming back with your views and ideas

**Comment by BendixSpring** [comment]
I’d like to throw my two cents in:

 

I think the user wants the right articles at the top of the list with the least amount of effort.

**Comment by BendixSpring** [comment]
I agree with Arvind.  But I do think we need to consider both a free form search (with operators) and some sort of checkbox or filter view.

 

Different people take different paths to get somewhere.  We should allow for that.

**Comment by BendixSpring** [comment]
This is probably a feature request, but if we are talking about making a better user experience, I think we need to consider functionality examples from successful sites.

 

For example, I’d like to have a list of my favorite searches.

Or my last N searches, so that I can go back and find what I was looking at 3 days ago.

Or here’s what other people ALSO looked at when they read article X (like Amazon)

 

This also leads into having a better user experience.  So if I am a Java coder, I’d like to be able to set a *system* filter to only show me stuff that is Java related.  I don’t want to set that every time I log in.  Give the user some preferences:  How things are displayed, how many, what is the arrangement of the panels, etc.  It makes it a more personal experience.

**Comment by DinisCruz** [comment]
@BendixSpring those are all valid requests which we aim to address with this new version

@roman87 can you confirm that all the ideas mentioned above (namely https://github.com/TeamMentor/TM_4_0_Design/issues/350#issuecomment-72479672) are captured in a issue and story? thx

**Comment by arvinddoraiswamy** [comment]
@DinisCruz : Agree completely on [comment 72420478](https://github.com/TeamMentor/TM_4_0_Design/issues/350#issuecomment-72420478). That's the best way to do it and we need to keep on continuously improving the efficiency of our search. Might take years.. but its the best way.

And in the meantime simplify the navigation with less categories as in [comment 72418137](https://github.com/TeamMentor/TM_4_0_Design/issues/350#issuecomment-72418137)

@BendixSpring : Agree that the user should not only get *articles* .. but the right articles at the top of the screen. Both those will be addressed automatically IMO, if the search and navigation improves.

**Comment by roman87** [comment]
This is a great discussion.

@BendixSpring I think what you are suggesting regarding saving searches and articles is tracked here: https://github.com/TeamMentor/Master/issues/945
The global filter is tracked here: https://github.com/TeamMentor/Master/issues/947
Can you double check if I missed anything? Obviously this would mean that the user has to be able to delete/edit his saved searches. But that is a different story :) 

@arvinddoraiswamy  Classification is a science and there is no correct way of doing it. For example, I think that language is a subset of vulnerability, not the other way, like your are thinking. This is just a different perspective. We need to cater to multiple perspectives. 
Searching for things is also a science. People used to go to school to be librarieans. When I went to school they taught us how to find things in a library. There are courses now that will teach people how to use google. 
Basically we cannot expect people to be born with the ability to use TM. However we need to make it as intuitive for them as possible so they don't have to go to a class on TM :). 
That is why I agree, search is the starting point. However, intuative tags and filters, allow for much faster results in a closed content space like ours.

**Comment by arvinddoraiswamy** [comment]
@roman87 Fair enough :). As long as we're thinking about how to make both search and the filters.. lean and mean... AND functional, its cool.

**Comment by BendixSpring** [comment]
Dinis:

 

You did miss something:  Having saved searches and articles is important, but I was also asking for some sort of list of the previous N searches, regardless of what they were.  This is analogous to the browser history.  Actually saving something requires work on the user’s part – that’s equivalent to manually saving a site to favorites.  I’m asking for something that is even less effort…

**Comment by roman87** [comment]
Ah.. good point. Added that to the https://github.com/TeamMentor/Master/issues/945

**Comment by roman87** [comment]
Looks like we nailed this issue. Going to wait for the next beta to make sure everyone is happy with the solution.

**Comment by roman87** [comment]
As it turns out this is not going to work in Jade. This will be fixed in Flare

**Comment by romichg** [comment]
@luchia do we still want to do this? or is our current solution addressing the problem?

**Comment by luchia** [comment]
I think it's a good idea to do it.

**Comment by luchia** [comment]
I've added the icons, this will be pushed in my PR today. I need Michael or Dinis to add the functionality.
